x-pack/plugins/observability_solution/observability_ai_assistant/public/service/complete.test.ts

```diff
@@ -0,0 +1,398 @@
+            ),
+            map((event) => event.message),
+            toArray(),
+            last()
+          )
+        );
+      });
+
+      it('calls the request callback again with the executed message', () => {
+        expect(requestCallback).toHaveBeenCalledTimes(2);
+
+        const nextMessages = requestCallback.mock.lastCall![0].params.body.messages;
+
+        const expectedMessages = [
+          {
+            '@timestamp': expect.any(String),
+            message: {
+              content: '',
+              function_call: {
+                arguments: JSON.stringify({ foo: 'bar' }),
+                name: 'my_action',
+                trigger: MessageRole.Assistant,
+              },
+              role: MessageRole.Assistant,
+            },
+          },
+          {
+            '@timestamp': expect.any(String),
+            message: {
+              content: JSON.stringify({ bar: 'foo' }),
+              name: 'my_action',
+              role: MessageRole.User,
+            },
+          },
+        ];
+
+        expect(nextMessages).toEqual([...messages, ...expectedMessages]);
+      });
+
+      it('calls the action handler with the arguments from the LLM', () => {
+        expect(respondFn).toHaveBeenCalledWith(
+          expect.objectContaining({
+            args: {
+              foo: 'bar',
+            },
+          })
+        );
+      });
+
+      it('returns all the messages in the created observable', () => {
+        expect(allMessages[allMessages.length - 1]).toEqual({
+          '@timestamp': expect.any(String),
+          message: {
+            content: 'Great action call',
+            function_call: {
+              arguments: '',
+              name: '',
+              trigger: MessageRole.Assistant,
+            },
+            role: MessageRole.Assistant,
+          },
+        });
+      });
+    });
+
+    describe('and it fails', () => {
+      beforeEach(async () => {
+        respondFn.mockRejectedValueOnce(new Error('foo'));
+
+        requestCallback.mockImplementationOnce(() =>
+          of(...createLlmResponse([{ content: 'Action call failed' }]))
+        );
+
+        await lastValueFrom(
+          callComplete({
+            getScreenContexts,
+          }).pipe(
+            filter(
+              (event): event is MessageAddEvent =>
+                event.type === StreamingChatResponseEventType.MessageAdd
+            ),
+            map((event) => event.message),
+            toArray(),
+            last()
+          )
+        );
+      });
+
+      it('calls the request callback again with the error', () => {
+        expect(requestCallback).toHaveBeenCalledTimes(2);
+
+        const nextMessages = requestCallback.mock.lastCall![0].params.body.messages;
+
+        expect(nextMessages[nextMessages.length - 1]).toEqual({
+          '@timestamp': expect.any(String),
+          message: {
+            content: JSON.stringify({
+              error: {},
+              message: 'foo',
```

Comment:
it's because we set message in `createFunctionResponseError`. `JSON.stringify({ error: new Error()})` results in an empty object. `JSON.stringify({ error: new CustomError() })` results in `{ customErrorProp: 'foo' }`. I'll make the serialization a bit more explicit though (and copy the name, msg, stack).